Add a tf.contrib.image.translate function

ringw opened the pull request:
Opening as a replacement for #10748. Made a small simplification. @sampepose

sampepose committed:
Add a tf.contrib.image.translate function

ringw committed:
Merge remote-tracking branch 'upstream/master' into translate

ringw committed:
Remove redundant checks from tf.contrib.image.translate.

ringw committed:
Add translate and translations_to_projective_transforms to the docstring.

vrv commented on tensorflow/contrib/image/python/ops/image_ops.py:
```diff
@@ -122,6 +146,39 @@ def angles_to_projective_transforms(angles, image_height, image_width):
       axis=1)
 
 
+def translations_to_projective_transforms(translations):
```
does this need to be public, or can it be an hidden?

ringw replied:
We already expose angles_to_projective_transforms so that users can compose transformation matrices, so my plan was to make similar public methods for any transformation.

ringw committed:
Fix lint errors for tf.contrib.image.translate

vrv reviewed (commented):
Letting eugene look at the rest of this.

sampepose commented:
Thanks, @ringw! Sorry I haven't gotten a chance to finish off that PR. Lgtm

ahundt commented:
@sampepose I think you need to sign the CLA before this can be merged. Otherwise it will be stuck forever. :-)

sampepose commented:
hmm I've signed the CLA, not sure why this isn't going through.

SP

On Sun, Sep 10, 2017 at 3:05 PM, Andrew Hundt <notifications@github.com>
wrote:

> @sampepose <https://github.com/sampepose> I think you need to sign the
> CLA before this can be merged. Otherwise it will be stuck forever. :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/12306#issuecomment-328368007>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABxRu0REWoqPGQVlVshEeWV3gOqQZ_e2ks5shEEZgaJpZM4O4H3z>
> .
>

ebrevdo commented on tensorflow/contrib/image/python/ops/image_ops.py:
```diff
@@ -81,6 +81,31 @@ def rotate(images, angles, interpolation="NEAREST"):
     return output
 
 
+def translate(images, translations, interpolation="NEAREST"):
```
name= arg on any function that creates new ops., use it inside a "with name_scope(...):" before creating any ops.  here, elsewhere.

ringw replied:
Added name= args and tf.name_scope(name) throughout.

ebrevdo commented on tensorflow/contrib/image/python/ops/image_ops.py:
```diff
@@ -122,6 +147,40 @@ def angles_to_projective_transforms(angles, image_height, image_width):
+  """
+  translation_or_translations = ops.convert_to_tensor(
+      translations, name="translations", dtype=dtypes.float32)
+  if len(translation_or_translations.get_shape()) == 1:
```
what if .get_shape() returns TensorShape(None)?

ringw replied:
That is an issue throughout, but it's really useful to do this validation eagerly and be able to return a tensor with the correct shape (HW, HWC, or NHWC). Added docstrings indicating which arguments need a static shape with a known rank.

ringw commented:
@ahundt @sampepose The googlebot comment says "The good news is that everyone that needs to sign a CLA (the pull request submitter and all commit authors) have done so. Everything is all good there.", so hopefully the CLA is OK?

ringw committed:
Add name_scopes in image_ops.

Indicate in the docstrings when the static shape of the arguments must
have a known rank.

ahundt commented:
Oh I see you're right, I had only looked at the `cla: no` label.

ringw committed:
Fix pyformat's weird docstring indentation.

drpngx commented:
@ringw we really need that CLA. Could you double check? Usually, it's a problem with the email address, or commits getting co-mingled the PR. Thanks.

ringw committed:
Merge remote-tracking branch 'upstream/master' into translate

ringw commented:
@drpngx The only details I see about the CLA are: "The bad news is that it appears that one or more commits were authored by someone other than the pull request submitter. We need to confirm that they're okay with their commits being contributed to this project. Please have them confirm that here in the pull request."

@sampepose confirmed in this pull request that he's OK with contributing his commit. Are there any further steps necessary? Sam's commit has sampepose@gmail.com which appears to be his email on his profile, and my commits have ringwalt@google.com. I don't see any other commits that got co-mingled.

ringw committed:
tf.name_scope -> ops.name_scope

drpngx commented:
@sampepose Please leave a comment that you're OK with it, and then we can override and merge.

However, we need an OK from @ebrevdo first. Thanks.

ringw commented:
Sam already replied with "LGTM".

drpngx commented:
@ringw I saw that, I just want a clear comment.

sampepose commented:
Looks good to me. Merge away.

> On Sep 21, 2017, at 10:55 AM, drpngx <notifications@github.com> wrote:
> 
> @ringw I saw that, I just want a clear comment.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sb2nov commented:
@drpngx is this okay to merge ?

drpngx commented:
I think we'd still like an approval from @ebrevdo. Once that is done, @gunan, @martinwicke or @jhseu can override the CLA flag and merge.

ebrevdo commented on tensorflow/contrib/image/python/kernel_tests/image_ops_test.py:
```diff
@@ -87,6 +87,22 @@ def test_rotate_odd(self):
                               [24, 19, 14, 9, 4]]])
 
+  def test_translate(self):
+    with self.test_session():
```
Move the new session inside the loop.

ringw replied:
Done.

ebrevdo commented on tensorflow/contrib/image/python/ops/image_ops.py:
```diff
@@ -37,16 +37,18 @@
 
 
-def rotate(images, angles, interpolation="NEAREST"):
+def rotate(images, angles, interpolation="NEAREST", name="rotate"):
```
name=None and use the default second argument to name_scope below

ringw replied:
Done throughout.

ebrevdo commented on tensorflow/contrib/image/python/ops/image_ops.py:
```diff
@@ -55,80 +57,155 @@ def rotate(images, angles, interpolation="NEAREST"):
+    image_or_images = ops.convert_to_tensor(images)
+    if image_or_images.dtype.base_dtype not in _IMAGE_DTYPES:
+      raise TypeError("Invalid dtype %s." % image_or_images.dtype)
+    if len(image_or_images.get_shape()) == 2:
```
Check that shape.ndims is not none first

ringw replied:
Done throughout.

ebrevdo commented on tensorflow/contrib/image/python/ops/image_ops.py:
```diff
@@ -55,80 +57,155 @@ def rotate(images, angles, interpolation="NEAREST"):
+        (num_rows, num_columns) (HW). The rank must be statically known (the
+        shape is not `TensorShape(None)`.
+    translations: A vector representing [dx, dy] or (if images has rank 4)
+        a vector of length num_images, with a [dx, dy] vector for each image in
```
You mean a matrix?

ringw replied:
Done.

ebrevdo commented on tensorflow/contrib/image/python/ops/image_ops.py:
```diff
@@ -55,80 +57,155 @@ def rotate(images, angles, interpolation="NEAREST"):
+  Raises:
+    TypeError: If `image` is an invalid type.
+  """
+  with ops.name_scope(name):
```
name, "translate"):

ringw replied:
Done throughout.

ebrevdo commented on tensorflow/contrib/image/python/ops/image_ops.py:
```diff
@@ -55,80 +57,155 @@ def rotate(images, angles, interpolation="NEAREST"):
+  with ops.name_scope(name):
+    angle_or_angles = ops.convert_to_tensor(
+        angles, name="angles", dtype=dtypes.float32)
+    if len(angle_or_angles.get_shape()) == 0:  # pylint: disable=g-explicit-length-test
```
Check ndims first

ringw replied:
Done.

ebrevdo commented on tensorflow/contrib/image/python/ops/image_ops.py:
```diff
@@ -55,80 +57,155 @@ def rotate(images, angles, interpolation="NEAREST"):
+    else:
+      raise TypeError("Translations should have rank 1 or 2.")
+    num_translations = array_ops.shape(translations)[0]
+    return array_ops.concat(
```
Explain what you're doing here. Match dtype to image dtype?

ringw replied:
Actually, this just processes the translation matrices, which are always float32. Added a comment to clarify this, and describe the translation matrix.

ringw committed:
Move the test session inside the _DTYPES loop.

ringw committed:
Use the default_name arg of name_scope.

ringw committed:
Check for ndims == None

ringw committed:
Fix translate docstring and add a comment.

ringw committed:
s/vector/matrix/ for the multiple translations.

ringw commented:
Ping @ebrevdo

ringw commented:
@drpngx @gunan Could we go ahead and override the CLA flag and merge? Thanks!

gunan commented:
Sorry for the delay, merged.

ringw commented:
No problem!
